crates/torii/core/src/executor.rs

```diff
@@ -0,0 +1,298 @@
+    Model as ModelRegistered,
+};
+
+pub(crate) const LOG_TARGET: &str = "torii_core::executor";
+
+#[derive(Debug, Clone)]
+pub enum Argument {
+    Null,
+    Int(i64),
+    Bool(bool),
+    String(String),
+    FieldElement(Felt),
+}
+
+#[derive(Debug, Clone)]
+pub enum BrokerMessage {
+    ModelRegistered(ModelRegistered),
+    EntityUpdated(EntityUpdated),
+    EventMessageUpdated(EventMessageUpdated),
+    EventEmitted(EventEmitted),
+}
+
+#[derive(Debug, Clone)]
+pub struct DeleteEntityQuery {
+    pub entity_id: String,
+    pub event_id: String,
+    pub block_timestamp: String,
+    pub ty: Ty,
+}
+
+#[derive(Debug, Clone)]
+pub enum QueryType {
+    SetEntity(Ty),
+    DeleteEntity(DeleteEntityQuery),
+    EventMessage(Ty),
+    RegisterModel,
+    StoreEvent,
+    Execute,
+    Other,
+}
+
+#[derive(Debug)]
+pub struct Executor<'c> {
+    pool: Pool<Sqlite>,
+    transaction: Transaction<'c, Sqlite>,
+    publish_queue: VecDeque<BrokerMessage>,
+    rx: UnboundedReceiver<QueryMessage>,
+    shutdown_rx: Receiver<()>,
+}
+
+#[derive(Debug)]
+pub struct QueryMessage {
+    pub statement: String,
+    pub arguments: Vec<Argument>,
+    pub query_type: QueryType,
+    tx: Option<oneshot::Sender<Result<()>>>,
+}
+
+impl QueryMessage {
+    pub fn new(statement: String, arguments: Vec<Argument>, query_type: QueryType) -> Self {
+        Self { statement, arguments, query_type, tx: None }
+    }
+
+    pub fn new_recv(
+        statement: String,
+        arguments: Vec<Argument>,
+        query_type: QueryType,
+    ) -> (Self, oneshot::Receiver<Result<()>>) {
+        let (tx, rx) = oneshot::channel();
+        (Self { statement, arguments, query_type, tx: Some(tx) }, rx)
+    }
+
+    pub fn other(statement: String, arguments: Vec<Argument>) -> Self {
+        Self { statement, arguments, query_type: QueryType::Other, tx: None }
+    }
+
+    pub fn other_recv(
+        statement: String,
+        arguments: Vec<Argument>,
+    ) -> (Self, oneshot::Receiver<Result<()>>) {
+        let (tx, rx) = oneshot::channel();
+        (Self { statement, arguments, query_type: QueryType::Other, tx: Some(tx) }, rx)
+    }
+
+    pub fn execute() -> Self {
+        Self {
+            statement: "".to_string(),
+            arguments: vec![],
+            query_type: QueryType::Execute,
+            tx: None,
+        }
+    }
+
+    pub fn execute_recv() -> (Self, oneshot::Receiver<Result<()>>) {
+        let (tx, rx) = oneshot::channel();
+        (
+            Self {
+                statement: "".to_string(),
+                arguments: vec![],
+                query_type: QueryType::Execute,
+                tx: Some(tx),
+            },
+            rx,
+        )
+    }
+}
+
+impl<'c> Executor<'c> {
+    pub async fn new(
+        pool: Pool<Sqlite>,
+        shutdown_tx: Sender<()>,
+    ) -> Result<(Self, UnboundedSender<QueryMessage>)> {
+        let (tx, rx) = unbounded_channel();
```

Comment:
i think we should use a bounded channel for the executor to apply back pressure, incase there are alot of complex queries and executor is slower compared to the processor. wdyt?

Reply:
The execute QueryMessage can kinda be used to apply backpressure. In this case we're executing at the end of the syncing to guarantee one Tx for everything but that can easily be refactored to spread queries into multiple Txs which should relieve the channel. 

I have experimented with a bounded channel, it might be beneficial in the future but for now it is just easier to have an unbounded one as we won't need to refactor all of our `sql.rs` ops to be asynchronous.